Mathlib/AlgebraicGeometry/Stalk.lean

```diff
@@ -51,12 +61,264 @@ theorem IsAffineOpen.fromSpecStalk_eq {X : Scheme} (x : X) {U V : X.Opens}
+
+/-- https://stacks.math.columbia.edu/tag/01J7 -/
+lemma range_fromSpecStalk {x : X} :
+    Set.range (X.fromSpecStalk x).1.base = { y | y ⤳ x } := by
+  ext y
+  constructor
+  · rintro ⟨y, rfl⟩
+    exact ((LocalRing.specializes_closedPoint y).map (X.fromSpecStalk x).1.base.2).trans
+      (specializes_of_eq fromSpecStalk_closedPoint)
+  · rintro (hy : y ⤳ x)
+    have := fromSpecStalk_closedPoint (x := y)
+    rw [← stalkSpecializes_fromSpecStalk hy] at this
+    exact ⟨_, this⟩
+
+end Scheme
+
+end fromSpecStalk
+
+variable (R : CommRingCat.{u}) [LocalRing R]
+
+section stalkClosedPointIso
+
+/-- For a local ring `(R, 𝔪)`,
+this is the isomorphism between the stalk of `Spec R` at `𝔪` and `R`. -/
+noncomputable
+def stalkClosedPointIso :
+    (Spec R).presheaf.stalk (closedPoint R) ≅ R :=
+  StructureSheaf.stalkIso _ _ ≪≫ (IsLocalization.atUnits R
+      (closedPoint R).asIdeal.primeCompl fun _ ↦ not_not.mp).toRingEquiv.toCommRingCatIso.symm
+
+lemma stalkClosedPointIso_inv :
+    (stalkClosedPointIso R).inv = StructureSheaf.toStalk R _ := by
+  ext x
+  exact StructureSheaf.localizationToStalk_of _ _ _
+
+lemma ΓSpecIso_hom_stalkClosedPointIso_inv :
+    (Scheme.ΓSpecIso R).hom ≫ (stalkClosedPointIso R).inv =
+      (Spec R).presheaf.germ ⊤ (closedPoint _) trivial := by
+  rw [stalkClosedPointIso_inv, ← Iso.eq_inv_comp]
+  rfl
+
+@[reassoc (attr := simp)]
+lemma germ_stalkClosedPointIso_hom :
+    (Spec R).presheaf.germ ⊤ (closedPoint _) trivial ≫ (stalkClosedPointIso R).hom =
+      (Scheme.ΓSpecIso R).hom := by
+  rw [← ΓSpecIso_hom_stalkClosedPointIso_inv, Category.assoc, Iso.inv_hom_id, Category.comp_id]
+
+lemma Spec_stalkClosedPointIso :
+    Spec.map (stalkClosedPointIso R).inv = (Spec R).fromSpecStalk (closedPoint R) := by
+  rw [stalkClosedPointIso_inv, Scheme.Spec_fromSpecStalk']
+
+end stalkClosedPointIso
+
+section stalkClosedPointTo
+
+variable {R} (f : Spec R ⟶ X)
+
+-- move me
+lemma LocalRing.closed_point_mem_iff {R : Type*} [CommRing R] [LocalRing R]
+    {U : Opens (PrimeSpectrum R)} : closedPoint R ∈ U ↔ U = ⊤ :=
+  ⟨(eq_top_iff.mpr fun x _ ↦ (specializes_closedPoint x).mem_open U.2 ·), (· ▸ trivial)⟩
+
+@[simp]
+lemma Spec_closedPoint {R S : CommRingCat} [LocalRing R] [LocalRing S]
+    {f : R ⟶ S} [IsLocalRingHom f] : (Spec.map f).1.base (closedPoint S) = closedPoint R :=
+  LocalRing.comap_closedPoint f
+
+namespace Scheme
+
+/--
+Given a local ring `(R, 𝔪)` and a morphism `f : Spec R ⟶ X`,
+they induce a ring homomorphim `φ : 𝒪_{X, f 𝔪} ⟶ X`.
+
+This is inverse to `φ ↦ Spec.map φ ≫ X.fromSpecStalk (f 𝔪)`. See `SpecToEquivOfLocalRing`.
+-/
+noncomputable
+def stalkClosedPointTo :
+    X.presheaf.stalk (f.1.base (closedPoint R)) ⟶ R :=
+  f.stalkMap (closedPoint R) ≫ (stalkClosedPointIso R).hom
+
+instance isLocalRingHom_stalkClosedPointTo :
+    IsLocalRingHom (stalkClosedPointTo f) := by
+  apply (config := { allowSynthFailures := true }) RingHom.isLocalRingHom_comp
+  · apply isLocalRingHom_of_iso
+  · apply f.2
```

Comment:
Do you understand why `infer_instance` does not work?

Reply:
I hope/believe #17403 will solve this.